### PR TITLE
TELCODOCS-1188 creating new PR about adding new parameter

### PR DESCRIPTION
**Pull request by kquinn1204**
[TELCODOCS-1002]:  [CNF-6356](https://issues.redhat.com//browse/CNF-6356) SR-IOV Network Operator alongside with host-reserved SR-IOV VFs

Version(s):
4.14, 4.15 and main

Issue:
https://issues.redhat.com/browse/TELCODOCS-1188

Link to docs preview: 
https://62354--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-sriov-device.html#nw-sriov-networknodepolicy-object_configuring-sriov-device

QE review:
- [x] QE has approved this change.

Additional information:
On hold and keeping for z stream release.

### CODE REVIEW

**Comment by kquinn1204** [comment]
@SchSeba Hi can you PTAL this PR for CNF-6356

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (SchSeba) [inline modules/nw-sriov-networknodepolicy-object.adoc:28]: externallyManage -> externallyManaged

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (SchSeba) [inline modules/nw-sriov-networknodepolicy-object.adoc:59]: maybe something like 

```
Allow the SR-IOV network operator to configure and allocate a subset of virtual functions from
a physical function that is configured externally from SR-IOV network operator.
```

please take a look on the design proposal in the sriov repo https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/476/files

**Comment by kquinn1204** [comment]
@SchSeba PTAL I have updated the description as per your guidance. I hope it is OK.

**Comment by kquinn1204** [comment]
@evgenLevin this will need a LGTM from QE so I would be grateful for a LGTM on this

**Comment by kquinn1204** [comment]
On hold and keeping for z stream release

**Review by SchSeba — COMMENTED** [review]
as always great work!

LGTM
but adding hold until the feature is merged

/hold

**Comment by kalexand-rh** [comment]
I've opened the `enterprise-4.15` branch. Please review your PR and update the applicable versions to also cover 4.15 if you expect that the changes in this PR will be valid in that branch.

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (evgenLevin) [inline modules/nw-sriov-networknodepolicy-object.adoc:72]: Please remove "test"

**Review by evgenLevin — COMMENTED** [review]
lgtm

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:72]: yep I just added that to try to trigger rebuild of preview. Will remove Otherwise is it OK ?

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc]: just removed. Thanks

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc]: OK see LGTM now thanks @evgenLevin

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (cgoncalves) [inline modules/nw-sriov-networknodepolicy-object.adoc:71]: The first sentence says the Operator configures a subset of VFs when `externallyManaged=true`. The second sentence says that the Operator configures the VFs when `externallyManaged=false`.

When `externallyManaged=true`, the Operator will not allocate or configure the VFs. The Operator can use the VFs to attach them to Pods but not manage them (i.e. create/allocate and configure)

**Review by cgoncalves — CHANGES_REQUESTED** [review]
@SchSeba @evgenLevin please consider my in-line comment and let us know if I am wrong.

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:71]: I thought the key point was that with this set to true the SR-IOV Network Operator is can configure a subset of "all" VF whereas with it set to false (think default) the SR-IOV Network Operator does not have that granular control and configures "all"

Comment (cgoncalves) [reply modules/nw-sriov-networknodepolicy-object.adoc:71]: On the contrary. The parameter is called `externallyManaged` which hints that VFs are not to be managed by the Operator. The default value is `false`, yes.

Comment (cgoncalves) [reply modules/nw-sriov-networknodepolicy-object.adoc:71]: Please consider the following suggestion:

"Set `externallyManaged` to `true` to allow the SR-IOV Network Operator to use all or a subset of externally managed virtual functions and attach them to Pods. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs."

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:71]: Thank you, @cgoncalves "to use **all** or a subset of externally managed virtual functions" seems like a important distinction which I must have missed.

Comment (evgenLevin) [reply modules/nw-sriov-networknodepolicy-object.adoc]: @cgoncalves  You are absolutely right:
* If externallyCreated is false (default) - SR-IOV must create/recreate the VFs on its own.
* if externallyCreated is true - SR-IOV should use VFs created earlier.

Comment (evgenLevin) [reply modules/nw-sriov-networknodepolicy-object.adoc]: @kquinn1204
 BTW This line is only relevant if externallyCreated is true. 

> Before applying the policy, it is necessary to create the VFs first; otherwise, the webhook will block the request.

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc]: @evgenLevin I made that clarification thanks

**Review by skrthomas — COMMENTED** [review]
/lgtm

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (SchSeba) [inline modules/nw-sriov-networknodepolicy-object.adoc:76]: so the SR-IOV Network Operator will not take any actions on the PF  and the VFs that are not in the policy nitSelector.

**Review by SchSeba — COMMENTED** [review]
just small comment and we should be good to go thanks for working on this!

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:76]: Hi @SchSeba is this feature now released. I was waiting for the z-stream. Is it out and in what z-stream ?

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:76]: HI @SchSeba should this read "so the SR-IOV Network Operator will not take any actions on the PF and the VFs that are not in the policy nodeSelector" ?

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:76]: @SchSeba got it I think "so the SR-IOV Network Operator will not take any actions on the PF and the VFs that are not defined in the policy nicSelector"

Comment (SchSeba) [reply modules/nw-sriov-networknodepolicy-object.adoc]: yep I think it sounds good thanks!

the backport to 4.14.z is WIP should be done this week

**Comment by kalexand-rh** [comment]
The `branch/enterprise-4.16` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.15. And any PR going into `main` must also target the latest version branch (enterprise-4.16).

If the update in your PR does NOT apply to version 4.16 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:75]: Which VFs, the ones used by the Host?
And which policy, please clarify we are specifically referring to the SriovNetworkNodePolicy resource
And finally, which webhook are we talking about and which request will it block?

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:76]: What does 'resetting` mean in this context?

Suggest to clarify that any VFs used on the host system must be created through NMState (not manually), while the SR-IOV operator will manage VFs presented as Pod secondary interfaces as always

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:79]:
```suggestion
<10> The NIC selector identifies the device for which this resource applies to. You do not have to specify values for all the parameters. It is recommended to identify the network device with enough precision to avoid selecting a device unintentionally.
```

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:83]:
```suggestion
<11> Optional: The vendor hexadecimal vendor identifier of the SR-IOV network device. The only allowed values are `8086` (Intel) and `15b3` (Mellanox).
```

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:85]:
```suggestion
<12> Optional: The device hexadecimal device identifier of the SR-IOV network device. For example, `101b` is the device ID for a Mellanox ConnectX-6 device.
```

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:87]:
```suggestion
<13> Optional: An array of one or more physical function (PF) names the resource should apply to.
```

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:89]:
```suggestion
<14> Optional: An array of one or more PCI bus addresses the resource should apply to.  
Example: `0000:02:00.1`.
```
A PF can not be spread across multiple PCI devices

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:91]: Please clarify (or link) what a network filter is

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:93]:
```suggestion
<16> Optional: The driver to configure for the VFs created from this resource. The only allowed values are `netdevice` and `vfio-pci`. The default value is `netdevice`.
```

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [inline modules/nw-sriov-networknodepolicy-object.adoc:109]: Please specify what type of hardware offloading is enabled

Comment (kquinn1204) [reply modules/nw-sriov-networknodepolicy-object.adoc:109]: Good question I did not directly write this piece do you know what type of hardware offloading is enabled ?

**Comment by kquinn1204** [comment]
Hi @djuran I did merge the changes with this PR https://github.com/openshift/openshift-docs/pull/72304 earlier in the week. I will kick off another PR with your updates above as thta would be cleaner I feel.

File: modules/nw-sriov-networknodepolicy-object.adoc
Comment (djuran) [reply modules/nw-sriov-networknodepolicy-object.adoc:109]: Not really and since this is not clear to me, I'm guessing it won't be clear to a customer either

**Pull request closed without merging**
